Add more tests for inventorydb funcs during vehicle/ECU registration tests

awwad opened the pull request:
To improve coverage (and also defend code from future changes), I've added a variety of calls to the inventorydb functions within code that tests the results of vehicle and ECU registration. That test code already checked the values internal to the inventorydb, but didn't always test the associated functions. It now does a better job of that.

awwad committed:
Also test inventorydb funcs during vehicle registration tests

awwad committed:
Move a few tests in ECU registration testing for tidiness

awwad committed:
Test inventorydb unknown ECU responses in ecu registration tests

awwad committed:
Test for correct Spoofing errors during ECU registration

awwad committed:
Add test for inventorydb.register_vehicle, including overwriting

vladimir-v-diaz commented on tests/test_director.py:
```diff
@@ -191,6 +191,10 @@ def test_03_add_new_vehicle(self):
 
+    # The VIN should be unknown to the inventory db.
+    with self.assertRaises(uptane.UnknownVehicle):
+      inventory.get_last_vehicle_manifest(vin)
```
Is there a reason why you chose `inventory` over `inventorydb`?  I find it easier to see that it's the `uptane.services.inventorydb` module that it represents.

awwad replied:
A good point. I think I should actually change the module's name to inventory. I'll do that in a future PR.

vladimir-v-diaz commented on tests/test_director.py:
```diff
@@ -200,6 +204,8 @@ def test_03_add_new_vehicle(self):
     self.assertIn(vin, inventory.ecus_by_vin)
     self.assertIn(vin, inventory.primary_ecus_by_vin)
+    # The VIN is now known, but there should be no Vehicle Manifests yet
+    self.assertIsNone(inventory.get_last_vehicle_manifest(vin))
```
I was tracing `get_last_vehicle_manifest()` and noticed this outdated LOC:
https://github.com/uptane/uptane/blob/821464acb89e33b9b65894e6da4b98272e74cca3/uptane/services/inventorydb.py#L386-L387

awwad replied:
Resolved.

vladimir-v-diaz commented on tests/test_director.py:
```diff
@@ -252,6 +252,26 @@ def test_05_register_ecu_serial(self):
+    self.assertFalse(inventory.ecu_public_keys)
+    self.assertFalse(inventory.vehicle_manifests[vin])
+    self.assertFalse(inventory.ecu_manifests)
+    self.assertIsNone(inventory.primary_ecus_by_vin[vin])
```
How `primary_ecus_by_vin[vin]` is verified to be empty stood out to me.  Is there a reason why None is used instead of a Boolean, unlike the others?  I didn't dig in deeply, but if all these either store dicts and lists, shouldn't you consistently check for booleanness?

awwad replied:
Each vehicle has a single entry in primary_ecus_by_vin; rather than a list, it's a string. If that vehicle has no registered Primary, the value is None.

vladimir-v-diaz commented on tests/test_director.py:
```diff
@@ -299,6 +329,29 @@ def test_05_register_ecu_serial(self):
+    self.assertFalse(inventory.get_ecu_manifests(secondary_serial))
+    self.assertIsNone(inventory.get_last_ecu_manifest(secondary_serial))
+
+    # Try registering the Secondary again. Normally, we expect a Spoofing error
```
Try to break up some of these lines, easier to read IMO.  You might have chunked all this together to make it a logical unit, but I found it difficult to eyeball.

vladimir-v-diaz commented on tests/test_director.py:
```diff
@@ -299,6 +329,29 @@ def test_05_register_ecu_serial(self):
+
+    # Try registering the Secondary again. Normally, we expect a Spoofing error
+    # as with the Primary, but for now, the registration attempt is simply
+    # ignored instead of raising an exception, for hacky demo reasons. /:
```
Is the Secondary code documented to make it clear that although a hacky implementation is used, it will eventually change?  Perhaps a reason why a hacky implementation is in place might help with maintainability.

awwad replied:
The problem originates with the fact that the external web frontend code incorrectly performs ECU registrations every time it assigns a target to an ECU. I think I'll just fix that now. I'll give it a few minutes of work and see how that goes.

awwad replied:
Fixed.

vladimir-v-diaz commented on tests/test_director.py:
```diff
@@ -694,6 +747,35 @@ def test_40_add_target_for_ecu(self):
 
 
 
+  def test_60_register_vehicle(self):
```
Consider testing for invalid vin arguments.  I noticed there is a `_check_registration_is_sane()` function, so you might as well test it.  
Also, why do you use an assert in `_check_registration_is_sane` rather than a specific exception?

awwad replied:
A FormatError is raised if the given VIN is not a the correct format. If, however, the invariants are broken on the data structure, that is a result of an error in the code (or, I suppose, someone tinkering directly with the contents of inventorydb), not an error related to any input or use of the class. That struck me as a candidate for use of an assertion rather than an exception. I'm not confident that that's the right policy, though. What do you think I should raise when the code is observed to be broken?

vladimir-v-diaz replied:
I thought `_check_registration_is_sane()` was validating the arguments.  I guess having a test for invalid arguments and checking against FormatError is fine.

awwad replied:
It does, but it also checks to see if situations that should be impossible have arisen to detect coding errors (e.g. an ostensibly registered VIN appears in only one of the X>1 places you'd expect a registered VIN to appear)

awwad replied:
But anyway, I'm still not sure asserts are the right way to do that. What do you think?

vladimir-v-diaz replied:
I recommend keeping the assert if it's truly an impossible to reach internal error, and separating the other validation checks (so that we can test for them via exceptions).  We should also limit the number of asserts used throughout the code, if possible.

awwad replied:
Thanks. I'll add a test for FormatError.

awwad committed:
Remove demo hack in inventorydb, raising SpoofingError correctly

The external web frontend used to re-register an ECU every time
it selected a new target file for it, which was unnecessary. I've
fixed that, and so now inventorydb.register_ecu can correctly
raise uptane.Spoofing if there is an attempt to re-register an
already registered ECU (without it being explicitly expected to
overwrite an existing registration). This also updates the test
in test_director to expect the correct behavior.

Also spaces out a few lines that Vlad found hard to eyeball
due to cramping, and clarifies a comment.

awwad committed:
Add some argument testing to inventorydb and test_directory

awwad committed:
Remove commented-out cruft in inventorydb

awwad commented:
Tests succeed, demo succeeds, review comments addressed. Merging. (Travis-CI is sometimes failing on 3.5 for reasons that have nothing to do with the project, but 3.3 and 3.4 are fine.)
